llarp/link/link_manager.hpp

```diff
@@ -0,0 +1,102 @@
+#ifndef LLARP_LINK_MANAGER_HPP
+#define LLARP_LINK_MANAGER_HPP
+
+#include <link/i_link_manager.hpp>
+
+#include <util/compare_ptr.hpp>
+#include <util/threading.hpp>
+#include <link/server.hpp>
+
+#include <unordered_map>
+#include <set>
+#include <atomic>
+
+namespace llarp
+{
+
+  struct IRouterContactManager;
+
+  struct LinkManager : public ILinkManager
+  {
+  public:
+
+    ~LinkManager() = default;
+
+    LinkLayer_ptr
+    GetCompatibleLink(const RouterContact &rc) const override;
+
+    IOutboundSessionMaker *
+    GetSessionMaker() const override;
+
+    bool
+    SendTo(const RouterID& remote, const llarp_buffer_t& buf) override;
+
+    bool
+    HasSessionTo(const RouterID& remote) const override;
+
+    void
+    PumpLinks() override;
+
+    void
+    AddLink(LinkLayer_ptr link, bool inbound = false) override;
+
+    bool
+    StartLinks(Logic_ptr logic) override;
+
+    void
+    Stop() override;
+
+    void
+    PersistSessionUntil(const RouterID &remote, llarp_time_t until) override;
+
+    void
+    ForEachPeer(std::function< void(const ILinkSession *, bool) > visit,
+                bool randomize = false) const override;
+
+    void
+    ForEachPeer(std::function< void(ILinkSession *) > visit) override;
+
+    void
+    ForEachInboundLink(std::function< void(LinkLayer_ptr) > visit) const override;
+
+    size_t
+    NumberOfConnectedRouters() const override;
+
+    size_t
+    NumberOfConnectedClients() const override;
+
+    bool
+    GetRandomConnectedRouter(RouterContact &router) const override;
+
+    void
+    CheckPersistingSessions(llarp_time_t now) override;
+
+    virtual util::StatusObject
+    ExtractStatus() const override;
+
+    void
+    Init(IOutboundSessionMaker *sessionMaker);
+
+  private:
+
+    LinkLayer_ptr
+    GetLinkWithSessionTo(const RouterID &remote) const;
+
+    std::atomic< bool > stopping;
+    mutable util::Mutex _mutex; // protects m_PersistingSessions
+
+    using LinkSet = std::set< LinkLayer_ptr, ComparePtr< LinkLayer_ptr > >;
+
+    LinkSet outboundLinks;
+    LinkSet inboundLinks;
+
+    // sessions to persist -> timestamp to end persist at
+    std::unordered_map< RouterID, llarp_time_t, RouterID::Hash >
+        m_PersistingSessions GUARDED_BY(_mutex);
+
+    IOutboundSessionMaker *_sessionMaker;
```

Comment:
should this be a `const *` ?

Reply:
only if I change it to be set in the constructor as @michael-loki suggested, but in case / because some of these things need to point to one another I opted to set all of their pointers in an Init function.  Naturally the pointers are private to at least mostly mitigate the risk of accidentally changing them rather than using them.

Reply:
I suppose I could make these pointers `const` and cast the `const` away in the `Init` function, but that also feels messy.  Thoughts?

Reply:
not casting const away, no change.